### ISSUE TITLE
Inference doesn't find common base type for union with null or undefined - TypeScript 4.8 regression

### ISSUE THREAD

**Issue by dragomirtitian** [microsoft/TypeScript#49938]
# Bug Report

### 🔎 Search Terms

inference common base type null undefined 

### 🕗 Version & Regression Information

- This changed between versions 4.7.3 and 4.8-beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/GYVwdgxgLglg9mABAUwI4gIYBsA8AVAPgAoMAuRPAGkQCNy8BKRAb0QF8AoDrZKRAN0QBeRGBBYsAQgDcXDgHp5iAPI0AVsmiIoATwAOyAM6IA7jCgALRBDgBbWwloZDybfqNddBxACFhLRGA4OHJDKAAnGDAAc3YOL1cAEX9WIJDEMMiY6Sdw8jFbGmRwuIUlAFFw8LgSqMQAFgA6AA4KAGVrDDAAcj5gKIATazsHJATEQBlyMobGgHYKU2dEcAHkfrBkIYAfXw40TCwiQSWfamPjFbWozcQdxIYuRUQ4AGsFmGM-QFByPfRsI8QJzOgOM92mr3en1uyzAq3WNx++3+518wJO0Mu8IGDwUACpcRxELiVJZiigAB4YWx6HiGQm4+SPJQABUitnMMH4RkBsMQWHMxWwdOmlWqtSQTVaeA6EC6vUCg2G9kc4ymTya8zwiwusKuG22ohAhWKvwOAKWBSK4TROrh1wNAAZoQBGaEAJhxTwhWo+huNJURf0OKMtxRtiEQTp2rp2HtkXrePuMoZKO0x9sQgbNIaNVvDfqtGN1WIYsiZiAAglUMDpjGZLErRk4XG4DMKnqKaog6pL2p0en1FTZlWN3JNphqFiYlun9dCiCnoZkotEGABtAC6puRIMQC9zZJ2y5i6434dnNx2Kc3nqU3u7xn3-qXERXp8z2+Du6fhaPr5Pm7hteG4PAmkJ7ouf5ZKum5Fnac5ZjuSw-oeGT-jBZ4CN+kFodB75psW9oPEAA)

### 💻 Code

```ts
function equal<T>(a: T, b: T) { }

let v = null!;

// Object types with common base types

type B = { foo: string }
type D = { foo: string; bar: number }

// Error in 4.8 TS can't find common type ❌
// 4.7 T was undefined | B
equal(v as B, v as undefined | D)

// ok T is B ✅
equal(v as B, v as D)
// ok T is B | undefined ✅
equal(v as B, v as B | undefined)
```

### 🙁 Actual behavior

In 4.8 TypeScript can't find common base type between `B` and `undefined | D`. In 4.7 this used to be `B | undefined`. 

This also happens for literal types and their respective base types as well as for arrays as shown in the 'Other Examples' in the playground link. I didn't include the code as the bug is the same, it's just different instances we found in our code.

### 🙂 Expected behavior

Typescript finds the type `B | undefined` for `T` as it did in 4.7.

**Comment by dragomirtitian** [comment]
Also reported in Bloomberg feedback https://github.com/microsoft/TypeScript/issues/49835

**Comment by jakebailey** [comment]
This also bisects to #49119.

**Comment by jakebailey** [comment]
Also interestingly, if you write `equal(v as D | undefined, v as B)` (so, flip the argument order), `T` is inferred as `D | undefined`.

I'm a little confused as to why the argument order matters here, but given that the inference if you swap the arguments in 4.7 is `D | undefined` and it's a type error, it's nothing new.

[Playground Link](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/GYVwdgxgLglg9mABAUwI4gIYBsA8AVAPgAoMAuRPAGkQCNy8BKRAb0QF8AoDrZKRAN0QBeRGBBYsAQgDcXDgHp5iAPI0AVsmiIoATwAOyAM6IA7jCgALRBDgBbWwloZDybfqNddBxACFhLRGA4OHJDKAAnGDAAc3YOL1cAEX9WIJDEMMiY6Sdw8jFbGmRwuIUlABYAOgB2clBIWEc0TFw-AB9EcAATZGAo5C7iMl9EDu7e-q7qOhGxsB6+sAGGcn44GC6yxCqADjrwaHgkZuwcHyHyH2nLlYF1zZOsIkFnX2oX43HFgdHERIYtgA9AD8XEU2xq+waRxQ6FOyTmC0mFz+vy+k2uqMREyWXVuaw2W12UMOTThuARnXmOIGKMp6NxmPp1O+eNW9w4j2eiFeDJ+HUS7x5xh8TCBoKJkMCB0ax3JZxRV1oN3ZhPBxOl0LJLQVJEumNFqs2nPJ3NeSo+fwB4JBYIqUvqpLlOvaVKRuMVaJZGOVszdNLZdzV9r2mqdsJdXvdtL1fr5U19rvj+PuAB0TS0zSKheaowHrUoQUA)

**Comment by jakebailey** [comment]
I pinned this down to this part of the PR: https://github.com/microsoft/TypeScript/pull/49119/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R21147

This part of the code doesn't consider the fact that any of the types could be unions that contain `null` or `undefined`, only types which are exactly `null` or `undefined`. I have a fix which correctly retains `null` and `undefined` which fixes this issue but doesn't make things consistent in argument order (so, same as 4.7). Currently verifying that this is right and that I can't also fix the other case too, since it seems incongruent with other inferences.

**Comment by jakebailey** [comment]
My fix doesn't seem to be the right one; we should probably discuss the intended inference behavior here.

**Comment by ahejlsberg** [comment]
Fixed by #50021.
